services/DeviceInformation/source/DeviceInformationService.cpp

```diff
@@ -0,0 +1,127 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2020 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ble-service-device-information/DeviceInformationService.h"
+
+#if BLE_FEATURE_GATT_SERVER
+
+ble_error_t DeviceInformationService::add_service(
+    BLE &ble,
+    const char *manufacturers_name,
+    const char *model_number,
+    const char *serial_number,
+    const char *hardware_revision,
+    const char *firmware_revision,
+    const char *software_revision,
+    const SystemId *system_id,
+    const RegulatoryCertDataList *cert_data_list,
+    const PnPID *pnp_id
+) {
+    /* all characteristics are optional so we need to build an array that contains the requested ones */
+    size_t param_index = 0;
+
+    const char* param_strings[6];
+    UUID param_string_uuids[6];
+
+    if (manufacturers_name) {
+        param_strings[param_index] = manufacturers_name;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_MANUFACTURER_NAME_STRING_CHAR;
+    }
+    if (model_number) {
+        param_strings[param_index] = model_number;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_MODEL_NUMBER_STRING_CHAR;
+    }
+    if (serial_number) {
+        param_strings[param_index] = serial_number;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_SERIAL_NUMBER_STRING_CHAR;
+    }
+    if (hardware_revision) {
+        param_strings[param_index] = hardware_revision;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_HARDWARE_REVISION_STRING_CHAR;
+    }
+    if (firmware_revision) {
+        param_strings[param_index] = firmware_revision;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_FIRMWARE_REVISION_STRING_CHAR;
+    }
+    if (software_revision) {
+        param_strings[param_index] = software_revision;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_SOFTWARE_REVISION_STRING_CHAR;
+    }
+
+    GattCharacteristic* param_chars[9];
```

Comment:
Do we need an array of pointers to the characteristics. Can we avoid the dynamic allocation ?

Reply:
`Do we need an array of pointers to the characteristics.`
Yes, that's how service is constructed.
`Can we avoid the dynamic allocation ?`
No, I can allocate on stack or statically, which one do you prefer?

Reply:
It make sense to use the stack when we can. If it's under 500bytes we we can allocate it on the stack.

Reply:
I don't know how big the characteristics are as they are dynamic so I don't know I'm going to blow your stack. Also if I use the stack I have to allocate all the potential characteristics even if you use one of them. I really don't see the point, this is gatt we're talking about here, dynamic allocation will be made even if I allocate on the stack since each characteristic will malloc the buffer for the data.

Reply:
Ok, no strong opinion here, we can keep it that way.